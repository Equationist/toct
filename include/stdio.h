/* stdio.h - Standard Input/Output Library for CompilerKit C Compiler */

#ifndef _STDIO_H
#define _STDIO_H

/* Standard I/O types */
typedef struct {
    int _fd;
    int _flags;
    char* _buffer;
} FILE;

/* Standard streams */
extern FILE* stdin;
extern FILE* stdout; 
extern FILE* stderr;

/* Standard I/O functions */
int printf(const char* format, ...);
int fprintf(FILE* stream, const char* format, ...);
int sprintf(char* str, const char* format, ...);
int snprintf(char* str, int size, const char* format, ...);

int scanf(const char* format, ...);
int fscanf(FILE* stream, const char* format, ...);
int sscanf(const char* str, const char* format, ...);

int putchar(int c);
int puts(const char* s);
int fputs(const char* s, FILE* stream);
int fputc(int c, FILE* stream);

int getchar(void);
char* gets(char* s);
char* fgets(char* s, int size, FILE* stream);
int fgetc(FILE* stream);
int ungetc(int c, FILE* stream);

FILE* fopen(const char* filename, const char* mode);
FILE* freopen(const char* filename, const char* mode, FILE* stream);
int fclose(FILE* stream);
int fflush(FILE* stream);

long ftell(FILE* stream);
int fseek(FILE* stream, long offset, int whence);
void rewind(FILE* stream);
int feof(FILE* stream);
int ferror(FILE* stream);
void clearerr(FILE* stream);

/* File position constants */
#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2

/* End of file constant */
#define EOF (-1)

/* Buffer size */
#define BUFSIZ 8192

/* Null pointer */
#ifndef NULL
#define NULL ((void*)0)
#endif

#endif /* _STDIO_H */