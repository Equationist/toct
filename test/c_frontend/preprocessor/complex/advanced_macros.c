/* Test advanced macro techniques */
#define EVAL(x) x
#define CAT(a, b) a##b
#define XCAT(a, b) CAT(a, b)

/* X-Macro pattern for code generation */
#define COLORS \
    X(RED, 1) \
    X(GREEN, 2) \
    X(BLUE, 3) \
    X(YELLOW, 4)

/* Generate enum */
typedef enum {
#define X(name, value) COLOR_##name = value,
    COLORS
#undef X
} color_t;

/* Generate string array */
const char* color_names[] = {
#define X(name, value) [COLOR_##name] = #name,
    COLORS
#undef X
};

/* Multi-level macro expansion */
#define INDIRECT(m, x) m(x)
#define MAKE_NAME(prefix) XCAT(prefix, _function)
#define DECLARE_FUNCTION(name) int name() { return 42; }

DECLARE_FUNCTION(MAKE_NAME(test))
DECLARE_FUNCTION(MAKE_NAME(demo))

#include <stdio.h>

int main() {
    printf("Advanced macro tests:\n");
    
    printf("Colors via X-macro:\n");
    printf("RED = %d (%s)\n", COLOR_RED, color_names[COLOR_RED]);
    printf("GREEN = %d (%s)\n", COLOR_GREEN, color_names[COLOR_GREEN]);
    printf("BLUE = %d (%s)\n", COLOR_BLUE, color_names[COLOR_BLUE]);
    printf("YELLOW = %d (%s)\n", COLOR_YELLOW, color_names[COLOR_YELLOW]);
    
    printf("\nGenerated functions:\n");
    printf("test_function() = %d\n", test_function());
    printf("demo_function() = %d\n", demo_function());
    
    return 0;
}