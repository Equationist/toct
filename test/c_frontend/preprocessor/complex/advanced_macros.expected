Advanced macro tests:
Colors via X-macro:
RED = 1 (RED)
GREEN = 2 (GREEN)
BLUE = 3 (BLUE)
YELLOW = 4 (YELLOW)

Generated functions:
test_function() = 42
demo_function() = 42